### PR TITLE
Silhoutte plot

### PR DESCRIPTION
**Pull request by neelasha23**
Add silhouette plot and score.

### CODE REVIEW

**Comment by idomic** [comment]
@neelasha23 tag us when ready for review.
I assume you took the tests module as is? If so try to think if we're missing something (mainly on the integration that didn't exist in the module).
Also, we'll need to add docs, like easy code snippets users can plug and play.

**Comment by neelasha23** [comment]
Yes I'll tag. I did add a test case for the Integration,  but I'll  check if anything more can be added.

**Comment by neelasha23** [comment]
@idomic @edublancas This is ready for review. I have added around 6 test cases apart from the scikit-plot tests, but still the coverage score seems to be stuck at 92%.  Do I need to add anymore code snippets apart from the 4 example scripts?

File: tests/test_silhouette_plot.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    assert out_ax.get_ylim() == (0.0, 250.0)
+
+
+def test_invalid_clusterer():
+    clf = DecisionTreeClassifier()
+    with pytest.raises(TypeError):
+        plot.silhouette_plot(X, clf)
+
+
+@patch('sklearn_evaluation.plot.clustering.silhouette_plot_from_results')
```
Comment (idomic) [inline tests/test_silhouette_plot.py:138]: Please use monkey patch instead of @patch.

**Review by idomic — CHANGES_REQUESTED** [review]
92% coverage is fine.
Overall looks good, we'd want to add a guide as part of this PR too.
Look here, this can use you as [a sample](https://sklearn-evaluation.readthedocs.io/en/latest/user_guide/compare.html).
We want something similar but for clustering, for users who aren't aware of this we'll add some explanations on when to use it how to use it and what for.

**Comment by neelasha23** [comment]
@idomic have added a .ipynb file for silhouette. Do I need to add guide for elbow curve in this PR or create a separate PR?
Haven't added a .rst file, I think it would be having similar content as the `clustering_evaluation.ipynb` file

**Comment by idomic** [comment]
@neelasha23 Please include it as part of this PR

**Comment by idomic** [comment]
I don't believe there's a way to render the docs locally (I haven't seen it in the readme/tasks).
@edublancas thoughts? Either we push it and check the guide or we add similar scripts to ploomber so devs can test locally

**Comment by edublancas** [comment]
there's a way to render docs locally: https://github.com/ploomber/sklearn-evaluation/tree/master/docs

but the instructions might be incomplete. so if you find any issues, let us know so we can help and please document any missing steps

**Comment by neelasha23** [comment]
@idomic  @edublancas 
have added test cases and docs for elbow curve as well. A couple of points that I want to highlight:
1. It looks like there's some inconsistency with the matplotlib `@image_comparison` functionality. There were two test cases which passed on my local but failed in the CI so I have removed the image comparison for only those 2 cases. Running the `override_failed_result_images.py` script also didnt help in these cases.  Maybe we should include our own image comparison utility like `yellowbricks` has.
2. While building the docs locally I faced this [issue](https://github.com/ploomber/sklearn-evaluation/issues/71). It might be because sklearn has a module by the same name and some conflict is happening. I renamed the module as `classification_report_skeval.py` and that solved the issue. However I haven't pushed this change as it might be an issue with my local setup. Let me know if I still to push this

**Comment by idomic** [comment]
@neelasha23 I see some conflict, please solve it.

File: src/sklearn_evaluation/plot/clustering.py [hunk 2]
```diff
@@ -141,3 +142,167 @@ def _clone_and_score_clusterer(clf, X, n_clusters):
...
+                    cmap='nipy_spectral',
+                    text_fontsize="medium",
+                    ax=None):
+    """Plots silhouette analysis of clusters provided.
+
+    Parameters
+    -----------
+
+    X : array-like, shape = [n_samples, n_features]:
+        Data to cluster, where n_samples is the number of samples and
```
Comment (idomic) [inline src/sklearn_evaluation/plot/clustering.py:161]: Data to cluster -> Cluster data

Comment (idomic) [reply src/sklearn_evaluation/plot/clustering.py:161]: Also missing the type within the array

**Review by idomic — COMMENTED** [review]
Looks good, please fix comments and conflict.
I still need to run the docs locally to check the guide.

**Comment by idomic** [comment]
For 1. we can open an issue about it and move on, I think as long as it works without those 2 tests it's fine.
For 2. I'll build the docs now locally and lets see if this continues.

**Comment by idomic** [comment]
@yafimvo please review this PR.

File: src/sklearn_evaluation/plot/clustering.py [hunk 3]
```diff
@@ -141,3 +143,168 @@ def _clone_and_score_clusterer(clf, X, n_clusters):
     clf = clone(clf)
     setattr(clf, 'n_clusters', n_clusters)
     return clf.fit(X).score(X), time.time() - start
+
+
+def silhouette_plot(X,
```
Comment (yafimvo) [inline src/sklearn_evaluation/plot/clustering.py:148]: We need to track the usage of `def silhouette_plot` and `def silhouette_plot_from_results`. We can do it with `@SKLearnEvaluationLogger.log(feature='plot')`

File: tests/test_clustering.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+@image_comparison(baseline_images=['silhouette_plot_cosine'],
+                  extensions=['png'],
+                  remove_text=False)
+def test_metric():
+    clf = KMeans()
+    plot.silhouette_plot(X, clf, range_n_clusters=[6], metric='cosine')
+
+
+def test_string_classes():
```
Comment (yafimvo) [inline tests/test_clustering.py:152]: This test generates the file `silhouette_plot_string_classes.png` in the root folder. I suggest saving this file in some dedicated directory for test artifacts that we can add to .gitignore.

**Review by yafimvo — CHANGES_REQUESTED** [review]
I tested it locally and it looks good. 

2 notes:
1. After running the test `test_string_classes` it generates the file `silhouette_plot_string_classes.png` in the root folder (then generates a new possible commit). I suggest saving this file in some dedicated directory for test artifacts that we can add to .gitignore.
2. We need to track the usage of `def silhouette_plot` and `def silhouette_plot_from_results`. We can do it with `@SKLearnEvaluationLogger.log(feature='plot')`

File: tests/test_clustering.py (hunk 4 above)
Comment (neelasha23) [reply tests/test_clustering.py:152]: This was a miss from my end. I have removed the saving logic.

File: src/sklearn_evaluation/plot/clustering.py (hunk 3 above)
Comment (neelasha23) [reply src/sklearn_evaluation/plot/clustering.py:150]: added

**Comment by yafimvo** [comment]
looking good to me
@idomic @neelasha23

**Comment by idomic** [comment]
Good points, on 1 we should use a fixture to save it into a tmp folder.

**Comment by idomic** [comment]
@neelasha23 tag me when this is done and we can merge.

**Comment by neelasha23** [comment]
@idomic  I have fixed @yafimvo 's review comments. On point 1 , it was a miss from my end. I forgot to remove the image save logic, I was just using it for testing purpose. Have removed the logic now. If the docs look okay this is ready

**Review by idomic — APPROVED** [review]
lgtm

**Comment by idomic** [comment]
Nice job!

**Comment by edublancas** [comment]
the docs are still missing. I see there is a reference to `user_guide/clustering_evaluation` but the notebook isn't there. also, the new plots do not appear on the plots list in the docs.

**Comment by idomic** [comment]
He's talking about the [readthedocs](https://sklearn-evaluation.readthedocs.io/en/latest/)

**Comment by neelasha23** [comment]
I'm able to see it when I build the docs in my local:
i think it will be visible if this [PR](https://github.com/ploomber/sklearn-evaluation/pull/85) is merged?
Screenshot 2022-11-29 at 11 32 05 PM

**Comment by neelasha23** [comment]
I still don't see the guide [here](https://sklearn-evaluation.readthedocs.io/en/latest/)

**Comment by edublancas** [comment]
try deleting the cache in your browser: https://sklearn-evaluation.readthedocs.io/en/latest/user_guide/clustering_evaluation.html

**Comment by neelasha23** [comment]
it worked now !

**Pull request merged**